### DISCUSSION TITLE
Any way to support streaming of response via SSE?

### DISCUSSION THREAD

**Discussion by MatthewLoffredo** [saloonphp/saloon#336] [Q&A]
Hey all, I'm trying to do something and a bit confused. How would I go about streaming the response of a request (similar to how many ai providers are doing it) using SSE or being able to execute a function callback on each stream event? I know it has something to do with $curl_info[CURLOPT_WRITEFUNCTION] but not sure how to synthesize this with Saloon. And advice would be appreciated!

**Comment by Sammyjo20** [comment]
Hey @MatthewLoffredo so I've had a quick look at the the openai-php/client library which I know uses Guzzle and has this streaming capability and it looks like they're just using Guzzle's stream option, which is really cool! You can do this really easily in Saloon.

In your connector/request, overwrite the `defaultConfig` method and add the following config option:


```php
protected function defaultConfig(): array
{
    return [
         'stream' => true,
    ];
}
```

This will instruct the API to return with a response right away but allow you to stream the response. After that, in Saloon you can access the underlying stream using:

```php
$stream = $response->stream();
```

Which you can then read X bytes at a time, outputting the text similar to how ChatGPT works. 

Let me know if this works!

**Reply by Sammyjo20** [reply]
Just wanted to add to this. Not sure of the best way to get each "chunk" of result back to the user. Maybe PHP has a way to stream back to the browser? Or maybe websockets?

You could do something like 

```php
$stream = $response->stream();
$chunks = 0;

while(! $stream->eof()) {
     broadcast(new OpenAIResponse($chunk++, $stream->read(5)));
}
```

Then on the front-end, you could listen to that websocket event and make sure you are rending the text in the order of "chunks" just case you receive them in the wrong order?
